### PR TITLE
Ledger lines clear stems when voices overlap

### PR DESCRIPTION
**Pull request by miiizen**
Resolves: #18724 

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

File: src/engraving/rendering/dev/chordlayout.cpp [hunk 1]
```diff
@@ -1823,6 +1823,34 @@ void ChordLayout::layoutChords1(LayoutContext& ctx, Segment* segment, staff_idx_
                 }
             }
 
+            // Create space between ledger lines and stems
+            // Above stave
+            Note* topUpNote = upStemNotes.back();
+            Note* bottomDownNote = downStemNotes.front();
+            if (!bottomUpNote->chord()->noStem() && bottomUpNote->chord()->stem()) {
```
Comment (cbjeukendrup) [inline src/engraving/rendering/dev/chordlayout.cpp:1830]: Is there a particular reason to check `noStem()`? As far as I know, whenever `noStem` is true, the stem is (supposed to be) removed, based on the value of `shouldHaveStem()`.

File: src/engraving/rendering/dev/chordlayout.cpp [hunk 2]
```diff
@@ -1823,6 +1823,34 @@ void ChordLayout::layoutChords1(LayoutContext& ctx, Segment* segment, staff_idx_
...
+                int bottomUpStemLen = bottomUpNote->chord()->stem()->length() / sp * 2;
+                int firstLedgerAbove = -2;
+                // Don't adjust unisons or 2nds as these move anyway
+                int sep = topUpNote->line() - bottomDownNote->line();
+                if (topDownNote->line() < -1 && topDownNote->line() < bottomUpNote->line() && sep > 1
+                    && bottomUpNote->line() - bottomUpStemLen <= firstLedgerAbove) {
+                    upOffset += 0.15 * sp;
```
Comment (cbjeukendrup) [inline src/engraving/rendering/dev/chordlayout.cpp:1837]: I see a hard-coded offset here, but the length of ledger lines can be varied using `Sid::ledgerLineLength`. Shouldn't that be taken into account?

File: src/engraving/rendering/dev/chordlayout.cpp [hunk 3]
```diff
@@ -1823,6 +1823,34 @@ void ChordLayout::layoutChords1(LayoutContext& ctx, Segment* segment, staff_idx_
...
+                }
+            }
+
+            // Below stave
+            if (!topDownNote->chord()->noStem() && topDownNote->chord()->stem()) {
+                int topDownStemLen = topDownNote->chord()->stem()->length() / sp * 2;
+                int firstLedgerBelow = staff->lines(bottomUpNote->tick()) * 2;
+                // Don't adjust unisons or 2nds as these move anyway
+                int sep = topUpNote->line() - bottomDownNote->line();
+                if (bottomUpNote->line() > staff->lines(bottomUpNote->tick()) * 2 - 1 && topDownNote->line() < bottomUpNote->line()
```
Comment (cbjeukendrup) [inline src/engraving/rendering/dev/chordlayout.cpp:1847]: What about using `firstLedgerBelow` here too?
```suggestion
                if (bottomUpNote->line() > firstLedgerBelow - 1 && topDownNote->line() < bottomUpNote->line()
```

**Comment by miiizen** [comment]
That review unearthed quite a few issues, apologies.  After a bit of a refactor:
- Moved code to a more correct place so we don't check separation twice
- Hooks are considered properly - 0.15sp created from the edge of the hook glyph to adjacent ledger line, and 0.15sp created when downstem hooks moved to right of upset note
- Made adjustment 0.15sp from the end of the ledger line, taking into account `Sid::ledgerLineLength`
- Differently sized noteheads taken into account
- Updated tests

**Review by cbjeukendrup — APPROVED** [review]

**Comment by cbjeukendrup** [comment]
@oktophonie This may need re-testing since there were quite some changes since your approval. But perhaps you already did, I'm not sure.

**Review by oktophonie — APPROVED** [review]

**Pull request merged**